### ISSUE TITLE
Maybe no bias in llama fc

### ISSUE THREAD

**Issue by KerwinKai** [NickL77/BaldEagle#3]
Hi, I am learning your code and I am a bit confused about whether the bias of the FC layer should follow the configuration in the config. json instead of always being true. For llama in official repo is no bias in the fc. I'm not sure if it will affect Eagle's performance.

https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/ea_model.py#L55

https://github.com/NickL77/BaldEagle/blob/master/train/modules/model/llama_eagle.py#L31C1-L31C9

**Comment by NickL77** [comment]
I think you bring up a good point. The default llama config has bias set to false and it appears the official repo sets bias based on that parameter.

My rationale for always setting `bias` to True because sglang loads the FC layer with the default value set to True: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/models/llama_eagle.py#L81
Linear documentation: https://pytorch.org/docs/stable/generated/torch.nn.Linear.html

I think this is a good question to bring up to the EAGLE authors, otherwise we can run ablations with and without bias to see if one is better. I am happy to match their implementation once we get word.

**Comment by KerwinKai** [comment]
I think the eval tool needs to align with the train tool when loading the model. For example, I trained the llama 3.1 model of bf16 using BladEagle, but Eagle loaded llama 3.1 with fp16. When I changed the Eagle code from fp16 to bf16, the acceleration effect of speculative sampling could be improved by 6%.

Also, I am confused about why Eagle uses FP16 to represent Llama3.1, which is a BF16 format model. Do you have any deeper insights?

**Comment by KerwinKai** [comment]
I try with and without bias in fc, It will indeed lead to a decrease in performance.

**Comment by NickL77** [comment]
I also noticed that loading in bfloat16 in sglang leads to a increase in improvement. I'm not sure why EAGLE uses FP16 instead of bfloat16. What is the test you did? Did you do this test on training or on inference?

Your test is that without bias, the performance decreases?

**Comment by KerwinKai** [comment]
Thanks for your reply. I am learning Eagle training with your BaldEagle repo. The training code looks much simpler than inference, and even tree_decode (https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/utils.py#L306) and other functions in the official repo are not included.

In this repo, we only need to adjust trainer.py, llama_eagle.py, and data.py to train my own eagle models with different structures. My understanding is that the goal of draft model training is to decode tokens that are as consistent as possible with the output of the base model based on the current token and previous tokens. So we only need to calculate the loss for the area covered by loss_mask.

And tree_decode is a decoding method proposed by Eagle to assist in draft model sampling. In other words, if there is another sample work similar to tree_decode, such as eagle2 vs eagle1, we can keep the draft model unchanged. If there are different draft models structures, we can keep tree_decode unchanged, such as eagle vs jakiro.

I don’t know if my understanding is correct. If you have time, please share your insights.

**Comment by NickL77** [comment]
> the goal of draft model training is to decode tokens that are as consistent as possible with the output of the base model based on the current token and previous tokens
Yes

> we only need to calculate the loss for the area covered by loss_mask
Yes, but the loss mask is to mask out user instructions. User instructions are not generated by the target model so they may be out of distribution.

In terms of draft tree and sampling, I just use sglang to run and check performance. Other than some parameters to tune, I don't anticipate needing to change training code, which is the focus of this repo. I have some sglang benchmarking scripts that I will release soon.

**Comment by KerwinKai** [comment]
I see. I also try to eval the accelerate use sglang. My device is H20 * 1, here are my configurations and observations:

**Service Launch Commands**  
- Without speculative sampling: 

  ```python3 -m sglang.launch_server --model-path /path/to/meta-llama/Llama-3.1-8B-Instruct --enable-torch-compile --disable-radix-cache --chunked-prefill-size -1 --disable-overlap-schedule --dtype float16```
- With speculative sampling（EAGLE2）:

    ```python3 -m sglang.launch_server --model /path/to/meta-llama/Llama-3.1-8B-Instruct --speculative-algorithm EAGLE --speculative-draft-model-path /path/to/EAGLE-LLaMA3.1-Instruct-8B --speculative-num-steps 5 --speculative-eagle-topk 8 --speculative-num-draft-tokens 32 --mem-fraction 0.80 --enable-torch-compile --torch-compile-max-bs 2 --disable-radix-cache --chunked-prefill-size -1 --disable-overlap-schedule --dtype float16```

**Test Commands**  
  ```python3 -m sglang.bench_serving --backend sglang --dataset-name sharegpt --num-prompts 100 --request-rate 2```


**Results**  
The Output token throughput（tok/s）of sglang is 416.39, the Output token throughput（tok/s）of sglang+eagle2 is 365.67
The Mean E2E（ms）of sglang is 1691.55, the Mean E2E（ms）of sglang+eagle2 is 6480.76
The Mean TTFT（ms）of sglang is 63.67, the  Mean TTFT（ms）of sglang+eagle2 is 103.80
The Mean ITL（ms）of sglang is 7.69, the Mean ITL（ms）of sglang+eagle2 is 30.14

The eagle2 test in official repo can achieve 2.6×

May be i ought to use send_one.py to test the performance.

**Comment by NickL77** [comment]
Yes, it's well known that speculative decoding becomes less effective as batch_size increases, which is why EAGLE papers report most of their experiments on batch_size=1.

**Comment by KerwinKai** [comment]
I think improving efficiency is an engineering optimization problem for the inference framework. Eagle can theoretically be compatible with other optimization methods of sglang. If Bladeagle needs any community contributions, I'd love to get involved.
